### PR TITLE
Structured Errors

### PR DESCRIPTION
**Pull request by ryansch**
Move structured errors to a tree structure.

Warning! This PR breaks the existing `errors` interface as it can no longer be treated as an enumerable!

Error messages are now available as `context.error_messages` or `context.errors.full_messages`.

A hash representation is available as `context.errors.to_h` for use with things like APIs.

## Release
This will be `v2.0.0`.

### CODE REVIEW

File: spec/metaractor_spec.rb [hunk 1]
```diff
@@ -256,13 +260,27 @@ def call
...
         expect(result.error_messages).to include(
-          '/user: Bad user'
+          'Invalid configuration',
+          'is_admin must be true or false',
+          'user.title cannot be blank',
+          'user.username must be unique',
+          'user.username must not be blank'
+        )
+
+        expect(result.errors.full_messages_for(:user)).to include(
```
Comment (morganick) [inline spec/metaractor_spec.rb]: I do like this.

Comment (ryansch) [reply spec/metaractor_spec.rb]: Yeah I lifted that from `ActiveModel::Errors`. It's a nice interface.

File: lib/metaractor/spec.rb [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+  module Spec
+    module Helpers
+      def context_creator(error_message: nil, error_messages: [], errors: [], valid: nil, invalid: nil, **attributes)
+        if error_message.present?
+          error_messages << error_message
+        end
+
+        result = Interactor::Context.build(attributes)
+        result.add_errors(messages: error_messages)
+        result.add_errors(errors: errors)
```
[the remark this answers is not in the document]
Comment (ryansch) [reply lib/metaractor/spec.rb]: You understand correctly! 🥇

File: lib/metaractor/spec.rb [hunk 3]
```diff
@@ -17,7 +17,10 @@ def context_creator(error_message: nil, error_messages: [], errors: [], valid: n
           result.invalidate!
         end
 
-        if !result.errors.empty? || (success != nil && !success) || (failure != nil && failure)
+        if !result.errors.empty? ||
+            result.invalid? ||
+            (success != nil && !success) ||
+            (failure != nil && failure)
```
Comment (morganick) [inline lib/metaractor/spec.rb]: It's weird to me that these are two different things in the interactor and one is not alias to the other.

Comment (ryansch) [reply lib/metaractor/spec.rb]: They basically function as aliases unless you do something crazy and specify them both.

**Pull request merged**